src/zpiv/zpivmodule.cpp

```diff
@@ -39,29 +86,40 @@ const uint256 PublicCoinSpend::signatureHash() const
 
 namespace ZPIVModule {
 
-    bool createInput(CTxIn &in, CZerocoinMint &mint, uint256 hashTxOut) {
-        libzerocoin::ZerocoinParams *params = Params().Zerocoin_Params(false);
-        uint8_t nVersion = mint.GetVersion();
-        if (nVersion < libzerocoin::PrivateCoin::PUBKEY_VERSION) {
-            // No v1 serials accepted anymore.
-            return error("%s: failed to set zPIV privkey mint version=%d\n", __func__, nVersion);
+    bool createInput(CTxIn &in, CZerocoinMint &mint, uint256 hashTxOut, const int spendVersion) {
+        bool fUseV1Params = mint.GetVersion() < libzerocoin::PrivateCoin::PUBKEY_VERSION;
+        libzerocoin::ZerocoinParams *params = Params().Zerocoin_Params(fUseV1Params);
+        if (fUseV1Params && spendVersion < PUBSPEND_SCHNORR) {
+            // v1 coins need at least version 4 spends
+            std::string errMsg = strprintf("Unable to create PublicCoinSpend version %d with coin version 1. "
+                    "Minimum spend version required: %d", spendVersion, PUBSPEND_SCHNORR);
+            return error("%s: %s", __func__, errMsg);
```

Comment:
I saw this code repeated few times, check if worth it to encapsulate it in another method and just do a: `if(!checkSpendVersion(fUseV1Params, spendVersion)) return error("whatever");` in every instance.

Reply:
There is actually only another place where we have both conditions joined together (in `createInput()`). 
Anyway less code repetition is always better. https://github.com/PIVX-Project/PIVX/pull/936/commits/d71e2235482b3a63af69679cda31470aebe1d00f adds an `isAllowed` method (both class and static version) to public spends.
Inside `validate()`, instead, imo it's better to keep the conditions separate to not have to re-enter a `if (version < PUBSPEND_SCHNORR)` block again.